server/src/main/scala/org/http4s/server/middleware/CSRFMiddleware.scala

```diff
@@ -0,0 +1,196 @@
+import org.http4s.server.Middleware
+import org.http4s.headers.{Cookie => HCookie}
+import org.http4s.util.{CaseInsensitiveString, NonEmptyList}
+
+import scalaz.{Kleisli, OptionT}
+import scalaz.concurrent.Task
+import scalaz.syntax.std.boolean._
+
+case class CSRFMiddleware(config: CSRFConfiguration,
+  clock: Clock = Clock.systemUTC()) {
+
+  /** Sign our token using the current time in milliseconds as a nonce
+    * Signing and generating a token is potentially a unsafe operation
+    * if constructed with a bad key.
+    */
+  protected[middleware] def signToken(token: String): Task[String] = {
+    val joined = token + "-" + clock.millis()
+    for {
+      instance <- Task.delay(Mac.getInstance(CSRFMiddleware.SigningAlgo))
+      _ <- Task.delay(instance.init(config.key))
+      out <- Task.delay(instance.doFinal(joined.getBytes("UTF-8")))
+    } yield joined + "-" + Base64.getEncoder.encodeToString(out)
+  }
+
+  /** Generate a new token **/
+  protected[middleware] def generateToken: Task[String] =
+    signToken(CSRFMiddleware.genTokenString)
+
+  /** Decode our CSRF token and extract the original token string to sign
+    */
+  protected[middleware] def extractRaw(token: String): OptionT[Task, String] =
+    token.split("-") match {
+      case Array(raw, nonce, signed) =>
+        val res = for {
+          instance <- Task.delay(Mac.getInstance(CSRFMiddleware.SigningAlgo))
+          _ <- Task.delay(instance.init(config.key))
+          out <- Task.delay(
+            instance.doFinal(
+              (raw + "-" + nonce).getBytes(StandardCharsets.UTF_8)))
+          r <- Task.point(
+            MessageDigest
+              .isEqual(out, Base64.getDecoder.decode(signed))
+              .option(raw))
+        } yield r
+        OptionT(res)
+      case _ =>
+        OptionT.none
+    }
+
+  /** Constructs a middleware that will check for the csrf token
+    * presence on both the proper cookie, and header values.
+    *
+    * If it is a valid token, it will then embed a new one,
+    * to effectively randomize the complete token while
+    * avoiding the generation of a new secure random Id, to guard
+    * against [BREACH](http://breachattack.com/)
+    *
+    */
+  def validate: Middleware[Request, MaybeResponse, Request, MaybeResponse] = {
+    service =>
+      Kleisli[Task, Request, MaybeResponse] { r =>
+        (for {
+          c1 <- CSRFMiddleware.cookieFromHeaders(r, config.cookieName)
+          c2 <- OptionT(
+            Task.point(r.headers.get(CaseInsensitiveString(config.headerName))))
+          raw1 <- extractRaw(c1.content)
+          raw2 <- extractRaw(c2.value)
+          response <- if (CSRFMiddleware.isEqual(raw1, raw2)) {
+            OptionT[Task, MaybeResponse](service(r).map(Some(_)))
+          } else {
+            OptionT.none[Task, MaybeResponse]
+          }
+          newToken <- OptionT[Task, String](signToken(raw1).map(Some(_)))
+        } yield
+          response.cata(
```

Comment:
I'm going to think a bit here on what happens with a `Pass`.  Have you thought about the desired semantics, or is this slinging code until it compiles to get some feedback?

Reply:
The `cata` is intentional. I _want it_ to 404 normally if it passes the csrf challenge, but return `Unauthorized` otherwise.